game-core/src/main/java/games/strategy/engine/data/changefactory/units/BombingUnitDamageChange.java

```diff
@@ -0,0 +1,71 @@
+package games.strategy.engine.data.changefactory.units;
+
+import games.strategy.engine.data.Change;
+import games.strategy.engine.data.GameData;
+import games.strategy.engine.data.Territory;
+import games.strategy.engine.data.Unit;
+import java.util.Collection;
+import java.util.UUID;
+import java.util.stream.Collectors;
+import org.triplea.java.collections.IntegerMap;
+
+/**
+ * A game data change that captures the damage caused to a collection of units by a bombing attack.
+ */
+public class BombingUnitDamageChange extends Change {
+  private static final long serialVersionUID = 7093658184880237574L;
+  private final IntegerMap<String> hits;
+  private final IntegerMap<String> undoHits;
+  private final Collection<String> territoriesToNotify;
+
+  private BombingUnitDamageChange(
+      final IntegerMap<String> hits,
+      final IntegerMap<String> undoHits,
+      final Collection<String> territoriesToNotify) {
+    this.hits = hits;
+    this.undoHits = undoHits;
+    this.territoriesToNotify = territoriesToNotify;
+  }
+
+  public BombingUnitDamageChange(
+      final IntegerMap<Unit> hits, final Collection<Territory> territoriesToNotify) {
+    this.hits = new IntegerMap<>();
+    this.undoHits = new IntegerMap<>();
+    hits.entrySet()
+        .forEach(
+            entry -> {
+              this.hits.put(entry.getKey().getId().toString(), entry.getValue());
+              this.undoHits.put(entry.getKey().getId().toString(), entry.getKey().getUnitDamage());
+            });
+    this.territoriesToNotify =
+        territoriesToNotify.stream().map(Territory::getName).collect(Collectors.toList());
+  }
+
+  @Override
+  protected void perform(final GameData data) {
+    hits.keySet()
+        .forEach(
+            unitId -> {
+              data.getUnits().get(UUID.fromString(unitId)).setUnitDamage(hits.getInt(unitId));
+            });
+    this.territoriesToNotify.forEach(
+        territory -> {
+          data.getMap().getTerritory(territory).notifyChanged();
+        });
+  }
+
+  @Override
+  public Change invert() {
+    return new BombingUnitDamageChange(undoHits, hits, territoriesToNotify);
+  }
+
+  @Override
+  public String toString() {
```

Comment:
Is the 'toString' required? Is it used anywhere to print text for users? If not, is it helpful?

Reply:
No idea.  Several change objects have a toString.  Several don't.

Reply:
A lot of 'toStrings' are included to help debugging in Eclipse that used to show the 'toString' of an object when in debug. That was before when you could inspect the values of an object and you'd see just the 'toString' value.

If you play a game with a break-point in this method you'll probably be able to determine if this is dead code relatively quickly. IMO we should not be porting dead-code into new code if we can help it. When we do a new release I don't think we'll want to go back and do any clean up beyond carefully deleting that which is marked for deletion.